YouTube.com: Only Audio Resumes after Pause, Video remains still

jyavenard opened the pull request:
#### 895987c5c5acc58caebdd74f48618bba50f8d2b0
<pre>
YouTube.com: Only Audio Resumes after Pause, Video remains still
<a href="https://bugs.webkit.org/show_bug.cgi?id=287123">https://bugs.webkit.org/show_bug.cgi?id=287123</a>
<a href="https://rdar.apple.com/144274075">rdar://144274075</a>

Reviewed by Youenn Fablet.

When the timebase&apos;s rate change from 0 to 1 (or vice versa) we can observe
that sometimes the currentTime moves slightly either backward or forward.

The VideoMediaSampleRenderer sets a timer to fire at the time the next
available video frame should be displayed.
When the timebase&apos;s rate change and the time moved forward slightly, the
timer doesn&apos;t fire.
This prevented from the next video frame to be painted and another timer to
be set for the next frame making the media playing only with audio and the
last painted video frame showing forever.

To get around this, we now observe the timebase rate change and when it
changes from 0, re-schedule a purge&amp;display cycle.
Code to observe the timebase&apos;s rate change is extracted from MediaPlayerPrivateMediaSourceAVFObjC
into its own separate EffectiveRateChangedListener class.

Manually tested. Automated testing is difficult as the typical way to check
that the video frames are regularly changing would be to paint them into a canvas.
However, painting into a canvas forcibly refresh the video frame showing.
Video frames can also be dropped if they are late, which can easily happen
on a busy host, making a reftest results inconsistent.

* Source/WebCore/SourcesCocoa.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/cocoa/EffectiveRateChangedListener.h: Added. Code was extracted from MediaPlayerPrivateMediaSourceAVFObjC
* Source/WebCore/platform/cocoa/EffectiveRateChangedListener.mm: Added.
(-[WebEffectiveRateChangedListenerObjCAdapter initWithEffectiveRateChangedListener:]):
(-[WebEffectiveRateChangedListenerObjCAdapter protectedListener]):
(WebCore::timebaseEffectiveRateChangedCallback):
(WebCore::EffectiveRateChangedListener::EffectiveRateChangedListener):
(WebCore::EffectiveRateChangedListener::~EffectiveRateChangedListener):
(WebCore::EffectiveRateChangedListener::effectiveRateChanged):
(WebCore::EffectiveRateChangedListener::stop):
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateAVFoundationObjC.h: Shuffle headers around to fix unified build.
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateAVFoundationObjC.mm:
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaSourceAVFObjC.mm:
(WebCore::MediaPlayerPrivateMediaSourceAVFObjC::MediaPlayerPrivateMediaSourceAVFObjC):
(WebCore::MediaPlayerPrivateMediaSourceAVFObjC::~MediaPlayerPrivateMediaSourceAVFObjC):
(-[WebEffectiveRateChangedListenerObjCAdapter initWithEffectiveRateChangedListener:]): Deleted.
(-[WebEffectiveRateChangedListenerObjCAdapter protectedListener]): Deleted.
(WebCore::EffectiveRateChangedListener::create): Deleted.
(WebCore::EffectiveRateChangedListener::effectiveRateChanged): Deleted.
(WebCore::timebaseEffectiveRateChangedCallback): Deleted.
(WebCore::EffectiveRateChangedListener::stop): Deleted.
(WebCore::EffectiveRateChangedListener::EffectiveRateChangedListener): Deleted.
* Source/WebCore/platform/graphics/cocoa/VideoMediaSampleRenderer.h:
* Source/WebCore/platform/graphics/cocoa/VideoMediaSampleRenderer.mm:
(WebCore::VideoMediaSampleRenderer::setTimebase):
(WebCore::VideoMediaSampleRenderer::clearTimebase):
* Source/WebCore/platform/mac/WebPlaybackControlsManager.mm:

Canonical link: <a href="https://commits.webkit.org/289922@main">https://commits.webkit.org/289922@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2089f76b728d746ebc2bcaf3f78a090f9dc9cd24

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/88438 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/7957 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/42875 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/93396 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/39192 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/90489 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/8344 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/16141 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/68209 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/25931 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/91440 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/6380 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/79992 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/48575 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/6152 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/34405 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/38300 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/76524 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/35291 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/95238 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/15613 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/11436 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/77074 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/15869 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/75848 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/76332 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/18782 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/20722 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/19141 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/8651 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/15629 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/20935 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/15370 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/18819 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/17152 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WebCore/platform/cocoa/EffectiveRateChangedListener.h:
```diff
@@ -0,0 +1,64 @@
+class EffectiveRateChangedListener;
+}
+
+@interface WebEffectiveRateChangedListenerObjCAdapter : NSObject
```
Can we move the declaration to the.mm file and only forward declare it via `OBJC_CLASS`?

jyavenard replied:
can try indeed.

jyavenard committed:
YouTube.com: Only Audio Resumes after Pause, Video remains still
https://bugs.webkit.org/show_bug.cgi?id=287123
rdar://144274075

Reviewed by Youenn Fablet.

When the timebase's rate change from 0 to 1 (or vice versa) we can observe
that sometimes the currentTime moves slightly either backward or forward.

The VideoMediaSampleRenderer sets a timer to fire at the time the next
available video frame should be displayed.
When the timebase's rate change and the time moved forward slightly, the
timer doesn't fire.
This prevented from the next video frame to be painted and another timer to
be set for the next frame making the media playing only with audio and the
last painted video frame showing forever.

To get around this, we now observe the timebase rate change and when it
changes from 0, re-schedule a purge&display cycle.
Code to observe the timebase's rate change is extracted from MediaPlayerPrivateMediaSourceAVFObjC
into its own separate EffectiveRateChangedListener class.

Manually tested. Automated testing is difficult as the typical way to check
that the video frames are regularly changing would be to paint them into a canvas.
However, painting into a canvas forcibly refresh the video frame showing.
Video frames can also be dropped if they are late, which can easily happen
on a busy host, making a reftest results inconsistent.

* Source/WebCore/SourcesCocoa.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/cocoa/EffectiveRateChangedListener.h: Added. Code was extracted from MediaPlayerPrivateMediaSourceAVFObjC
* Source/WebCore/platform/cocoa/EffectiveRateChangedListener.mm: Added.
(-[WebEffectiveRateChangedListenerObjCAdapter initWithEffectiveRateChangedListener:]):
(-[WebEffectiveRateChangedListenerObjCAdapter protectedListener]):
(WebCore::timebaseEffectiveRateChangedCallback):
(WebCore::EffectiveRateChangedListener::EffectiveRateChangedListener):
(WebCore::EffectiveRateChangedListener::~EffectiveRateChangedListener):
(WebCore::EffectiveRateChangedListener::effectiveRateChanged):
(WebCore::EffectiveRateChangedListener::stop):
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateAVFoundationObjC.h: Shuffle headers around to fix unified build.
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateAVFoundationObjC.mm:
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaSourceAVFObjC.mm:
(WebCore::MediaPlayerPrivateMediaSourceAVFObjC::MediaPlayerPrivateMediaSourceAVFObjC):
(WebCore::MediaPlayerPrivateMediaSourceAVFObjC::~MediaPlayerPrivateMediaSourceAVFObjC):
(-[WebEffectiveRateChangedListenerObjCAdapter initWithEffectiveRateChangedListener:]): Deleted.
(-[WebEffectiveRateChangedListenerObjCAdapter protectedListener]): Deleted.
(WebCore::EffectiveRateChangedListener::create): Deleted.
(WebCore::EffectiveRateChangedListener::effectiveRateChanged): Deleted.
(WebCore::timebaseEffectiveRateChangedCallback): Deleted.
(WebCore::EffectiveRateChangedListener::stop): Deleted.
(WebCore::EffectiveRateChangedListener::EffectiveRateChangedListener): Deleted.
* Source/WebCore/platform/graphics/cocoa/VideoMediaSampleRenderer.h:
* Source/WebCore/platform/graphics/cocoa/VideoMediaSampleRenderer.mm:
(WebCore::VideoMediaSampleRenderer::setTimebase):
(WebCore::VideoMediaSampleRenderer::clearTimebase):
* Source/WebCore/platform/mac/WebPlaybackControlsManager.mm:

Canonical link: https://commits.webkit.org/289922@main

eric-carlson commented on Source/WebCore/platform/cocoa/EffectiveRateChangedListener.mm:
```diff
@@ -0,0 +1,101 @@
+{
+    assertIsMainThread();
+    ASSERT(timebase);
+    // Observer removed MediaPlayerPrivateMediaSourceAVFObjC destructor.
```
This comment seems out of place, the class doesn't know anything about MediaPlayerPrivateMediaSourceAVFObjC

jyavenard replied:
it was copied from the original code. I should have removed it. too late here
